source/code/plugins/oms_diag.rb

```diff
@@ -0,0 +1,135 @@
+module OMS
+  class Diag
+
+    # Defaults for diagnostic
+    DEFAULT_IPNAME = "Diagnostics"
+    DEFAULT_TAG    = "diag.oms"
+
+    # Mandatory property keys
+    DI_KEY_LOGMESSAGE = 'LogData'
+    DI_KEY_IPNAME     = 'IPName'
+    DI_KEY_TYPE       = 'type'
+    DI_KEY_TIME       = 'time'
+    DI_KEY_AGENTGUID  = 'sourceHealthServiceId'
+
+    # Constants for dataitem values
+    DI_TYPE_XML   = 'System.PropertyBagData'
+    DI_TYPE_JSON  = 'JsonData'
+
+    # Record keys
+    RECORD_DATAITEMS  = 'DataItems'
+    RECORD_IPNAME     = 'IPName'
+    RECORD_MGID       = 'ManagementGroupId'
+
+    # Record constant values
+    RECORD_MGID_VALUE = '{00000000-0000-0000-0000-000000000002}'
+
+    class << self
+
+      # Method to be used by INPUT and FILTER plugins for logging
+
+      # Method:
+      # LogDiag(logMessage, tag=DEFAULT_TAG)
+      #
+      # Description:
+      # This is to be utilized for logging to the diagnostic
+      # channel.
+      #
+      # Parameters:
+      # @logMessage[mandatory]: The log message string to be logged
+      # @tag[optional]: The tag with which to emit the diagnostic log. The
+      # default value would be DEFAULT_TAG
+      # @ipname[optional]: IPName can be optionally provided to depict a
+      # customized one other than the DEFAULT_IPNAME in diagnostic event
+      # @properties[optional]: Hash corresponding to key value pairs that
+      # would be added as part of this data item.
+      #
+      # NOTE: Certain mandatory properties to the dataitem are added by default
+      def LogDiag(logMessage, tag=DEFAULT_TAG, ipname=DEFAULT_IPNAME, properties=nil)
+        # Process default values for tag and ipname if they are passed as nil
+        tag ||= DEFAULT_TAG
+        ipname ||= DEFAULT_IPNAME
```

Comment:
Well the intention is the user can do the following:

1. LogDiag('Hi') where tag is DEFAULT_TAG, ipname is DEFAULT_IPNAME, properties are nil
2. LogDiag('Hi', nil, SomeIpName) the tag is DEFAULT_TAG, ipname is SomeIpName, properties are nil
3. LogDiag('Hi', nil, nil, SomeProperties) the tag is DEFAULT_TAG, ipname is DEFAULT_IPNAME and properties are SomeProperties

The intention is not to raise errors while reporting diagnostic logs. The intention is to allow the user to define any single parameter of the three that is intended and the remaining should utilize the default values.